app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt

```diff
@@ -0,0 +1,126 @@
+                    val mapType = object : TypeToken<Map<String, Any>>() {}.type
+                    val data = gson.fromJson<Map<String, Any>>(json, mapType)
+
+                    val docRef = firestore.collection("users/$uid/$collectionName").document()
+                    batch.set(docRef, data)
+                }
+
+                batch.commit().await()
+            }.onFailure {
+                logger.e(it) { "Error while saving observations" }
+            }
+        }
+    }
+
+    private fun getCollectionNameForObservation(observation: Observation): String {
+        return when {
+            observation.code.coding.any { it.code == bodyWeightObservation.code } -> bodyWeightObservation.name
+            observation.code.coding.any { it.code == bloodPressureCollection.code } -> bloodPressureCollection.name
+            observation.code.coding.any { it.code == heartRateCollection.code } -> heartRateCollection.name
+            else -> {
+                logger.w { "Unknown observation code" }
+                throw UnsupportedOperationException("Unknown observation code")
+            }
+        }
+    }
+
+    private suspend fun <T> getObservation(collection: ObservationCollection): Result<T?> {
+        return withContext(ioDispatcher) {
+            runCatching {
+                val uid = firebaseAuth.currentUser?.uid
+                    ?: throw IllegalStateException("User not authenticated")
+                firestore.collection("users/$uid/${collection.name}")
+                    .orderBy("effectiveDateTime", Query.Direction.DESCENDING)
+                    .limit(1)
+                    .get()
+                    .await()
+                    .documents
+                    .firstOrNull()
+                    ?.let { document ->
+                        val json = document.data
+                        val jsonString = gson.toJson(json)
+                        jsonParser.parseResource(Observation::class.java, jsonString)
+                            .toRecord() as T
+                    }
+            }.onFailure {
+                logger.e(it) { "Error while getting latest ${collection.name} observation" }
+                null
+            }
+        }
+    }
+
+    private fun Observation.toRecord(): androidx.health.connect.client.records.Record {
+        return observationToRecordMapper.map(this)
+    }
+
+    suspend fun getLatestBloodPressureObservation(): Result<BloodPressureRecord?> {
+        return getObservation<BloodPressureRecord>(bloodPressureCollection)
+    }
+
+    suspend fun getLatestBodyWeightObservation(): Result<WeightRecord?> {
+        return getObservation<WeightRecord>(bodyWeightObservation)
+    }
+
+    suspend fun getLatestHeartRateObservation(): Result<HeartRateRecord?> {
+        return getObservation<HeartRateRecord>(heartRateCollection)
+    }
+}
+
+data class ObservationCollection(val name: String, val code: String)
```

Comment:
Can be marked as private and instead of code: String, I would use the proposed enum for `Loinc` instead, type-safety ftw